File: typescript/src/data-structures/singly-linked-list.ts
/**
 * Singly linked list implementation.
 * Time: O(1) front ops, O(n) back ops and search
 * Space: O(n)
 */

interface SinglyNode<T> {
  value: T;
  next: SinglyNode<T> | null;
}

export class SinglyLinkedList<T> {
  private head: SinglyNode<T> | null = null;
  private tail: SinglyNode<T> | null = null;
  private _size: number = 0;

  pushFront(value: T): void {
    const node: SinglyNode<T> = { value, next: this.head };
    this.head = node;
    if (!this.tail) {
      this.tail = node;
    }
    this._size++;
  }

  pushBack(value: T): void {
    const node: SinglyNode<T> = { value, next: null };
    if (this.tail) {
      this.tail.next = node;
    } else {
      this.head = node;
    }
    this.tail = node;
    this._size++;
  }

  popFront(): T | undefined {
    if (!this.head) {
      return undefined;
    }
    const value = this.head.value;
    this.head = this.head.next;
    if (!this.head) {
      this.tail = null;
    }
    this._size--;
    return value;
  }

  peekFront(): T | undefined {
    return this.head?.value;
  }

  peekBack(): T | undefined {
    return this.tail?.value;
  }

  contains(value: T): boolean {
    let current = this.head;
    while (current) {
      if (current.value === value) {
        return true;
      }
      current = current.next;
    }
    return false;
  }

  remove(value: T): boolean {
    if (!this.head) {
      return false;
    }

    if (this.head.value === value) {
      this.head = this.head.next;
      if (!this.head) {
        this.tail = null;
      }
      this._size--;
      return true;
    }

    let current = this.head;
    while (current.next) {
      if (current.next.value === value) {
        if (current.next === this.tail) {
          this.tail = current;
        }
        current.next = current.next.next;
        this._size--;
        return true;
      }
      current = current.next;
    }
    return false;
  }

  isEmpty(): boolean {
    return this._size === 0;
  }

  size(): number {
    return this._size;
  }

  clear(): void {
    this.head = null;
    this.tail = null;
    this._size = 0;
  }

  toArray(): T[] {
    const result: T[] = [];
    let current = this.head;
    while (current) {
      result.push(current.value);
      current = current.next;
    }
    return result;
  }
}

File: java/src/main/java/com/dsa/datastructures/SinglyLinkedList.java
package com.dsa.datastructures;

import java.util.Optional;

/**
 * A singly linked list implementation.
 * Time: O(1) for pushFront, O(n) for pushBack, search
 * Space: O(n)
 */
public class SinglyLinkedList<T> {
    private static class Node<T> {
        T value;
        Node<T> next;

        Node(T value) {
            this.value = value;
            this.next = null;
        }
    }

    private Node<T> head;
    private int size;

    public SinglyLinkedList() {
        this.head = null;
        this.size = 0;
    }

    public void pushFront(T value) {
        Node<T> node = new Node<>(value);
        node.next = head;
        head = node;
        size++;
    }

    public Optional<T> popFront() {
        if (head == null) {
            return Optional.empty();
        }
        T value = head.value;
        head = head.next;
        size--;
        return Optional.of(value);
    }

    public void pushBack(T value) {
        Node<T> node = new Node<>(value);
        if (head == null) {
            head = node;
        } else {
            Node<T> current = head;
            while (current.next != null) {
                current = current.next;
            }
            current.next = node;
        }
        size++;
    }

    public Optional<T> peekFront() {
        if (head == null) {
            return Optional.empty();
        }
        return Optional.of(head.value);
    }

    public boolean contains(T value) {
        Node<T> current = head;
        while (current != null) {
            if (current.value.equals(value)) {
                return true;
            }
            current = current.next;
        }
        return false;
    }

    public boolean isEmpty() {
        return head == null;
    }

    public int size() {
        return size;
    }

    public void clear() {
        head = null;
        size = 0;
    }
}
